sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -178,6 +179,46 @@ class Analyzer(
     }
   }
 
+  /**
+   * Replaces queries of the form "SELECT expr FROM A GROUP BY 1 ORDER BY 1"
+   * with a query of the form "SELECT expr FROM A GROUP BY expr ORDER BY expr"
+   */
+  object ResolveNumericReferences extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        val newGroups = groups.map {
+          case group if group.isInstanceOf[Literal] && group.dataType.isInstanceOf[IntegralType] =>
+            aggs(group.toString.toInt - 1) match {
+              case u: UnresolvedAlias =>
+                u.child match {
+                  case UnresolvedStar(_) => // Can't replace literal with column yet
+                    group
+                  case _ => u.child
+                }
+              case a: Alias => a.child
+              case a: AttributeReference => a
+            }
+          case group => group
+        }
+        Aggregate(newGroups, aggs, child)
+      case Sort(ordering, global, child) =>
+        val newOrdering = ordering.map {
+          case o if o.child.isInstanceOf[Literal] && o.dataType.isInstanceOf[IntegralType] =>
+            val newExpr = child.asInstanceOf[Project].projectList(o.child.toString.toInt - 1)
```

Comment:
Tests are currently failing to due this cast. A ```Sort```'s child doesn't have to be a ```Project```, use the LogicalPlan ```output``` attributes to get the proper attribute.